contrib/babelfishpg_tsql/src/pl_handler.c

```diff
@@ -4088,20 +4088,30 @@ _PG_fini(void)
  */
 
 static void
-terminate_batch(bool send_error, bool compile_error)
+terminate_batch(bool send_error, bool compile_error, int SPI_depth)
 {
 	bool		error_mapping_failed = false;
 	int			rc;
+	int 		current_spi_stack_depth;
 
 	HOLD_INTERRUPTS();
 
 	elog(DEBUG2, "TSQL TXN finish current batch, error : %d compilation error : %d", send_error, compile_error);
 
 	/*
 	 * Disconnect from SPI manager
+	 * Also cleanup remnant SPI connections
+	 * Ideally current depth should be same as 
+	 * when caller was connecting to SPI Manager
 	 */
-	if ((rc = SPI_finish()) != SPI_OK_FINISH)
-		elog(ERROR, "SPI_finish failed: %s", SPI_result_code_string(rc));
+	SPI_get_depth(&current_spi_stack_depth);
+	
+	if (current_spi_stack_depth < SPI_depth)
+		elog(FATAL, "SPI stack corrupted");
```

Comment:
we should add more details to error message like expected connection count and current count. Also, message should be "SPI connection stack is inconsistent"